tb/cocotb/soc/periph_fw/gen_periph_hex.py

```
@@ -0,0 +1,420 @@
+# ---------------------------------------------------------------------------
+# Firmware builder
+# ---------------------------------------------------------------------------
+def build_firmware(asm: Assembler):
+    """Populate assembler with the self-checking firmware.
+
+    All branch/jump targets use thunks so they resolve after labels are placed.
+    """
+    # ── Prologue: materialise base registers ─────────────────────────────────
+    # x2 = 0x0000_2000  (SRAM base)
+    l2, a2 = lui_addi(2, 0x0000_2000)
+    asm.emit(l2); asm.emit(a2)
```

Flake8 (7.3.0)

[error] multiple statements on one line (semicolon)

(E702)